.changeset/khaki-icons-trade.md

```diff
@@ -0,0 +1,22 @@
+---
+'@backstage/backend-common': patch
+---
+
+1. URL Reader's `readTree` method now returns a `sha` in the response along with the blob. The SHA belongs to the latest commit on the target. `readTree` also takes an optional `sha` in its options and throws a `NotModifiedError` if the SHA matches with the latest commit SHA on the url's target. This can be used in building a cache when working with URL Reader.
```

Comment:
Wondering about the choice of wording, "sha". That would be an implementation detail, does it make more sense to call it

- "hash" if we know that it's always going to be a hash, or
- "commitHash" if we know that it's always going to be a commit hash out of a version control system, or
- "etag" if we want it to be generic and always work for any kind of source? <--- this feels like the most opaque one

Reply:
Agreed. I would also prefer something else over `SHA` which hides the implementation detail, but isn't too vague about its meaning - So I like `hash` over `sha` and `etag`.

As of now, the `hash` means the hash of the commit. But in future, if URL Reader is reading from a blob storage for example, the storage would not have any commit details. So, the `hash` could be the checksum of the tree being returned.

Reply:
etag is what http uses though, so it's pretty established...

Reply:
intentionally hiding the underlying implementation

Reply:
Aah okay, thanks! TIL

Reply:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-None-Match

Reply:
After reading about ETag and If-None-Match approach, I now feel that it's exactly what we are doing here. 😁